lib/webdriveragent.js

```diff
@@ -244,215 +238,97 @@ class WebDriverAgent {
         }
       });
 
-      this.deviceLogs.on('exit', (code) => {
-        let msg = `${this.realDevice ? 'System' : 'Simulator'} log exited with code '${code}'`;
-        log.info(msg);
-        if (code) {
-          return reject(msg);
+      return (async () => {
+        try {
+          await this.xcodebuild.start();
+          await this.waitForStart();
+          resolve();
+        } catch (err) {
+          let msg = `Unable to start WebDriverAgent: ${err}`;
+          log.error(msg);
+          reject(msg);
+          return;
         }
-      });
-
-      try {
-        let startTime = await this.getStartTime();
-        await this.xcodebuild.start();
-        let agentUrl = await this.waitForStart(startTime);
-        resolve(agentUrl);
-      } catch (err) {
-        let msg = `Unable to start WebDriverAgent: ${err}`;
-        log.error(msg);
-        return reject(msg);
-      }
+      })();
     });
   }
 
-  async waitForStart (startTime) {
-    // we have to wait for the sim to start before we can tail the log file
-    if (!this.realDevice) {
-      await systemLogExists(this.device);
-    }
-
-    let agentUrl;
-    let lineCount = 0;
-    let showWaitingMessage = true; // turn off logging once we have hit the end
-
-    let startDetector = (stdout) => {
-      let match = AGENT_STARTED_REGEX.exec(stdout);
-      if (match) {
-        if (this.realDevice) {
-          // on a real device there may already be system logs that need to be
-          // passed before we get to the real startup logs, otherwise
-          // it will be "done" before booting
-          // Expect a line like:
-          //     Dec  7 10:55:32 iamPhone XCTRunner(WebDriverAgentLib)[386] <Notice>: ServerURLHere->http://localhost:8100<-ServerURLHere
-          // need to get the date and add the year from the start date
-          let dateString = `${stdout.substr(0, 6)} ${startTime.getFullYear()} ${stdout.substr(7, 8)}`;
-          let buildTime = new Date(dateString);
-          if (!buildTime.isAfter(startTime)) {
-            return false;
-          }
+  async waitForStart () {
+    // try to connect once every 0.5 seconds, until `wdaLaunchTimeout` is up
+    log.debug(`Waiting ${this.wdaLaunchTimeout}ms for WebDriverAgent to start`);
+    let curlArgs = [
+      '-X', 'GET',
+      '-H', 'Content-Type: application/json;charset=UTF-8, accept: application/json',
+      `http://localhost:${this.url.port}/status`
+    ];
+    try {
+      let retries = parseInt(this.wdaLaunchTimeout / 500, 10);
+      await retryInterval(retries, 500, async () => {
+        if (this.xcodebuild.processExited) {
+          // there has been an error elsewhere and we need to short-circuit
+          return;
         }
-
-        agentUrl = match[1];
-        if (!agentUrl) {
-          log.errorAndThrow(new Error('No url detected from WebDriverAgent'));
+        try {
+          let {stdout} = await exec('curl', curlArgs);
```

Comment:
Executing terminal commands is usually slower than using built-in instruments. How about sending HEAD http request with timeout to the WDA URL and checking the response code?